### ISSUE TITLE
Digital Ocean credential exposure leading to unwanted VMs created

### ISSUE THREAD

**Issue by dduportal** [jenkins-infra/helpdesk#2884]
## What happened?

- On Sunday 10th of April 2022, at 19:12 UTC, we received automated emails from Digital Ocean notifying us about a set of individual virtual machines freshly started and ready to use, while we only use Kubernetes auto-scaling on Digital Ocean.

- At 19:15 UTC, one of the administrators notified the rest of the team and we identified that his identity was used for creating these machines.

- At 19:17 UTC, we decided to take immediate actions to prevent any propagation:
  - This administrator was removed from the following services: Digital Ocean, GitHub (as GitHub SSO is used for authentication to Digital Ocean), Azure, AWS and Jenkins LDAP  
  - Digital Ocean cluster was disabled on ci.jenkins.io to prevent any suspicious usage, and the automated management (Terraform and Kubernetes Management) were disabled.
  - All the cloud resources on Digital Ocean were deleted (including the Kubernetes cluster as it is stateless) except one of the suspicious VMs that was diagnosed and then stopped (to keep data for audit later)
  - Our administrator rotated all of his personal credentials, and we deleted the Digital Ocean token

- At 20:00 UTC, we confirmed that no further suspicious activity was detected on Digital Ocean, neither on our other cloud services
  - The IP used for creating the machines was not found anywhere on the platform (audit logs, access logs, VPN access)
  - The administrator which identity was used did not see any traces of connections on any services of the platform (except Digital Ocean of course) in the past 2 days
  - No data tampering was found in the GitHub repositories of the organization

- 2 day laters (12th of April) we received a detailed audit log from DigitalOcean Security team: they confirmed that the virtual machines were created through an API Personal Access Token 
  - No access to GitHub was detected, even in read
  - This PAT leak is similar to https://github.com/jenkins-infra/helpdesk/issues/2834 

- 4 days later (14th of April), we weren’t able to find the exact source of the leak.
  - Despite the strong gut feeling about the personal token being leaked through GitHub checks, we were not able to find the token value in the GitHub checks through the GitHub API.
  - We cannot be sure if the token was leaked from the infrastructure, or from a personal device.

## What are the impacts?

- Overall cost for the project is ~ 75$ in the DigitalOcean bill.
  - Around 20 droplets (VMs) were running at 100% CPU in Digital Ocean
  - Once all resources were deleted, 1 VM was kept in a stopped state for audit log, for around 60h hours

- The Kubernetes workload was moved (and billed) on the AWS EKS cluster for the time being (automatic, as this is multi-cloud)

- The 10 builds of ci.jenkins.io which ran in Digital Ocean during the 10 minutes (10 pods) were only plugin builds: only “read” operation (e.g. no write, no credential in these builds).

- No GitHub (sensitive) data could have been accessed by misusing this token.

## How to avoid this from happening again?

- [x] Disable GitHub checks output (only keep the “green/yellow/red” usual status) at all by default on private instance
  - As reminder, we already constrained the output in https://github.com/jenkins-infra/kubernetes-management/pull/2103
  - But with the (more) recent https://github.com/jenkins-infra/kubernetes-management/pull/2175 change, we are able to disable github checks at all
  - So making it disabled by default in the helm-chart in charge of generating the job-dsl configuration sounds clearly the way to go. 
  - We’ll enable it per job if required.

- Increase safety layers in Digital Ocean setup:
  - [x] Use a technical user for Digital Ocean who owns the API tokens. It has MFA enabled of course (backup code encoded in sops). It avoids risking one of our admins accounts (and would help to find leaks in the future).
  - [x] Add API token expiration, to ensure it is rotated regularly (90 days? Less ? More ?)

- Ensure that credentials are restricted to the job using it 
  - [x] Already done since the AWS key exposure

- [x] Re-create a new DigitalOcean cluster for ci.jenkins.io

## References

- AWS credential exposure (similar to this one) in Jenkins Infra: https://github.com/jenkins-infra/helpdesk/issues/2834

**Comment by dduportal** [comment]
References:
- https://github.com/jenkins-infra/helm-charts/pull/120
- https://github.com/jenkins-infra/helm-charts/releases/tag/jenkins-jobs-0.3.0
- https://github.com/jenkins-infra/kubernetes-management/pull/2245

**Comment by dduportal** [comment]
Adding a new digitalocean cluster, created with the new credentials:
- Creation with terraform: https://github.com/jenkins-infra/digitalocean/pull/28

Please note that the cluster creation fails with the following error:

```
[](https://infra.ci.jenkins.io/blue/organizations/jenkins/terraform-jobs%2Fdigitalocean/detail/main/22/pipeline/71/#step-72-log-188)[](https://infra.ci.jenkins.io/blue/organizations/jenkins/terraform-jobs%2Fdigitalocean/detail/main/22/pipeline/71/#step-72-log-189)╷
│ Error: Error creating Kubernetes cluster: POST https://api.digitalocean.com/v2/kubernetes/clusters[](https://infra.ci.jenkins.io/blue/organizations/jenkins/terraform-jobs%2Fdigitalocean/detail/main/22/pipeline/71/#step-72-log-190)[](https://infra.ci.jenkins.io/blue/organizations/jenkins/terraform-jobs%2Fdigitalocean/detail/main/22/pipeline/71/#step-72-log-191)[](https://infra.ci.jenkins.io/blue/organizations/jenkins/terraform-jobs%2Fdigitalocean/detail/main/22/pipeline/71/#step-72-log-192)[](https://infra.ci.jenkins.io/blue/organizations/jenkins/terraform-jobs%2Fdigitalocean/detail/main/22/pipeline/71/#step-72-log-193)[](https://infra.ci.jenkins.io/blue/organizations/jenkins/terraform-jobs%2Fdigitalocean/detail/main/22/pipeline/71/#step-72-log-194): 422 (request "0f39e0bd-7548-4cad-b828-23123a17a8e4") validation error: invalid version slug
│ 
│   with digitalocean_kubernetes_cluster.doks_cluster,
│   on doks-cluster.tf line 1, in resource "digitalocean_kubernetes_cluster" "doks_cluster":
│    1: resource "digitalocean_kubernetes_cluster" "doks_cluster" {
│
```

Blocked by #2866 (no more 1.20 cluster)

**Comment by dduportal** [comment]
Closing as the Digital Ocean is now fully operationnal again.
